cpp/src/arrow/types/string.cc

```diff
@@ -61,6 +61,15 @@ Status StringArray::Validate() const {
   return BinaryArray::Validate();
 }
 
-TypePtr BinaryBuilder::value_type_ = TypePtr(new UInt8Type());
+// This used to be a static member variable of BinaryBuilder, but it can cause
+// valgrind to report a (spurious?) memory leak when needed in other shared
+// libraries. The problem came up while adding explicit visibility to libarrow
+// and libarrow_parquet
+static TypePtr kBinaryValueType = TypePtr(new UInt8Type());
```

Comment:
@emkornfield this was a pretty thorny issue that took me a while to figure out (beyond my C++ DSO knowledge). The issue was that the static `value_type_` was being used at compile time in translation units in `libarrow_parquet`, so something about the order of static initialization was causing a memory leak, or at least something valgrind thought was a memory leak. The problem didn't appear until symbol visibility became explicit